libfoundation/include/system-library.h

```diff
@@ -0,0 +1,170 @@
+
+/* An opaque custom value ref type representing a loaded library. */
+typedef struct __MCSLibrary *MCSLibraryRef;
+
+/* The binding to the internal MCTypeInfoRef for the MCSLibraryRef type */
+MC_DLLEXPORT MCTypeInfoRef
+MCSLibraryTypeInfo(void) ATTRIBUTE_PURE;
+
+/* ================================================================
+ * Construction and querying
+ * ================================================================ */
+
+/* Create an MCSLibraryRef object by loading the library from the specified
+ * path.
+ *
+ * If the path is absolute, then only an attempt to load that specific library
+ * will be made. If the path is relative, however, an attempt to load the
+ * library will use the system's default search order.
+ *
+ * On Mac, bundles, frameworks and dylibs are supported. In order to load a
+ * bundle, an absolute path must be supplied. Otherwise, if the path ends in
+ * <LEAF>.framwork then the path will be modified to end in
+ * <LEAF>.framework/<LEAF>. In the dylib and framework case, dlopen's default
+ * search strategy is used.
+ *
+ * On Windows, dlls are supported. They are loaded using LoadLibraryEx with
+ * the LOAD_WITH_ALTERED_SEARCH_PATH flag set.
+ *
+ * On Linux, shared objects (.so) are supported. They are loaded using dlopen's
+ * search strategy.
+ *
+ * On Android, shared objects (.so) are supported. They are loaded using
+ * dlopen's search strategy which requires that non-system libraries be loaded
+ * using absolute paths (otherwise they can fail to load).
+ *
+ * On iOS, only statically linked and registered objects are supported. The
+ * name is matched exactly with the name specified in the static registration
+ * record. */
+MC_DLLEXPORT bool
+MCSLibraryCreateWithPath(MCStringRef p_path,
+                         MCSLibraryRef& r_library);
+
+/* Create an MCSLibraryRef object by referencing the library currently loaded
+ * at the specified address. */
+MC_DLLEXPORT bool
+MCSLibraryCreateWithAddress(void *p_address,
+                            MCSLibraryRef& r_library);
+
+/* Copy the full native path to the specified library. If the library is of
+ * static type, this returns the path of the loadable object which the library
+ * is linked into. */
+MC_DLLEXPORT bool
+MCSLibraryCopyPath(MCSLibraryRef p_library,
+                   MCStringRef& r_path);
+
+/* Lookup the symbol in the specified library. */
+MC_DLLEXPORT void *
+MCSLibraryLookupSymbol(MCSLibraryRef p_library,
+                       MCStringRef p_symbol);
+
+#ifdef __MCS_INTERNAL_API__
+
+bool
+__MCSLibraryThrowCreateWithNativePathFailed(MCStringRef p_native_path);
+bool
+__MCSLibraryThrowCreateWithAddressFailed(void *p_address);
+bool
+__MCSLibraryThrowResolvePathFailed(void);
+
+#endif
+
+/* ================================================================
+ * Static binding (iOS only)
+ * ================================================================ */
+
+#ifdef __IOS__
+
+/* New style (C++ based) - not implemented yet */
+
+#if 0
```

Comment:
There are bunch of `#if 0`-ed out blocks in this PR.  It would make it more obvious how they're related if you used descriptive-but-undefined macros, e.g.

```c++
#if defined(MCS_LIBRARY_CXX_IOS_STATIC)
/* stuff */
#endif
```

There are a couple of places where you have to thing for a while to realize that the `#if 0` isn't just dead code that should be removed, but actually relates to the new style iOS static bindings.